src/clj/puppetlabs/services/jruby_pool_manager/impl/jruby_internal.clj

```diff
@@ -86,6 +86,29 @@
    pool :- jruby-schemas/pool-queue-type]
   (.borrowItemWithTimeout pool timeout TimeUnit/MILLISECONDS))
 
+(schema/defn insert-shutdown-poison-pill
+  [pool :- jruby-schemas/pool-queue-type]
+  (.insertPill pool (ShutdownPoisonPill. pool)))
+
+(schema/defn insert-poison-pill
+  [error :- Throwable
+   pool :- jruby-schemas/pool-queue-type]
+  (.insertPill pool (PoisonPill. error)))
+
+(schema/defn fill-pool-with-poison-pills!
+  "Fill an empty pool with poison pills. Should always be run on the
+  modify-instance-agent. insert-fn should be a fn that inserts a poison
+  pill of some kind into the pool"
+  [pool-state :- jruby-schemas/PoolState
+   insert-fn :- IFn]
+  (let [pool (:pool pool-state)
+        pool-size (:size pool-state)]
+    (when (not (= pool-size (.remainingCapacity pool)))
```

Comment:
If I'm reading your comment right, there might be some confusion.

The poison pill will only be inserted if there's an error that caused the flush (or an error during the flush), not if it's just a normal flush from the http api.

If it's a normal flush, and filesync tries to take the lock, it will block until the pool is back up. Since there's now only one pool, and there's now swapping of pools happening, we don't have to insert the retry pill in the old pool just in case someone has a reference to it.

If there's been some error and a pill has been inserted, then the server will be shutting down, and I imagine filesync will want to know about that at some point, whether it's through an exception during the `.lock` call or when it finds a `ShutdownPoisonPill` after a borrow